### PR TITLE
adding OLMoE architecture

### PR DESCRIPTION
**Pull request by Goekdeniz-Guelmez**
```text
python -m mlx_lm.generate --model allenai/OLMoE-1B-7B-0125-Instruct --prompt "what's your name"
Fetching 9 files: 100%|███████████████████████████████████████████████████████████| 9/9 [00:00<00:00, 44992.53it/s]
==========
I'm OLMo, an open language model developed by AI2. If you have any questions or need assistance, feel free to ask OLMo!
==========
Prompt: 17 tokens, 20.405 tokens-per-sec
Generation: 31 tokens, 2.770 tokens-per-sec
Peak memory: 13.885 GB
```

### CODE REVIEW

**Comment by Goekdeniz-Guelmez** [comment]
```text
python -m mlx_lm.lora \
--model allenai/OLMoE-1B-7B-0125-Instruct \
--train \
--data /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/MLX/data_smoll \
--fine-tune-type lora \
--num-layers 1 \
--batch-size 1 \
--iters 5 \
--val-batches 1 \
--steps-per-report 1 \
--steps-per-eval 100 \
--adapter-path /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/MLX/test_olmoe \
--save-every 100 \
--max-seq-length 12 \
--grad-checkpoint
Loading pretrained model
Fetching 9 files: 100%|██████████████████████████████████████████████████████████| 9/9 [00:00<00:00, 170039.35it/s]
Loading datasets
Training
Trainable parameters: 0.001% (0.082M/6919.162M)
Starting training..., iters: 5
[WARNING] Some sequences are longer than 12 tokens. The longest sentence 840 will be truncated to 12. Consider pre-splitting your data to save memory.
[WARNING] Some sequences are longer than 12 tokens. The longest sentence 1979 will be truncated to 12. Consider pre-splitting your data to save memory.
Iter 1: Val loss 6.026, Val took 2.052s
Iter 1: Train loss 6.026, Learning Rate 1.000e-05, It/sec 0.546, Tokens/sec 6.006, Trained Tokens 11, Peak mem 14.302 GB
[WARNING] Some sequences are longer than 12 tokens. The longest sentence 3838 will be truncated to 12. Consider pre-splitting your data to save memory.
Iter 2: Train loss 6.010, Learning Rate 1.000e-05, It/sec 0.561, Tokens/sec 6.175, Trained Tokens 22, Peak mem 14.302 GB
[WARNING] Some sequences are longer than 12 tokens. The longest sentence 2158 will be truncated to 12. Consider pre-splitting your data to save memory.
Iter 3: Train loss 5.993, Learning Rate 1.000e-05, It/sec 0.547, Tokens/sec 6.017, Trained Tokens 33, Peak mem 14.302 GB
[WARNING] Some sequences are longer than 12 tokens. The longest sentence 478 will be truncated to 12. Consider pre-splitting your data to save memory.
Iter 4: Train loss 5.976, Learning Rate 1.000e-05, It/sec 0.511, Tokens/sec 5.623, Trained Tokens 44, Peak mem 14.302 GB
[WARNING] Some sequences are longer than 12 tokens. The longest sentence 2246 will be truncated to 12. Consider pre-splitting your data to save memory.
[WARNING] Some sequences are longer than 12 tokens. The longest sentence 1661 will be truncated to 12. Consider pre-splitting your data to save memory.
Iter 5: Val loss 5.950, Val took 0.539s
Iter 5: Train loss 5.950, Learning Rate 1.000e-05, It/sec 0.571, Tokens/sec 6.284, Trained Tokens 55, Peak mem 14.302 GB
Saved final weights to /Users/gokdenizgulmez/Library/Mobile Documents/com~apple~CloudDocs/Datastes/MLX/test_olmoe/adapters.safetensors.
```

**Comment by Goekdeniz-Guelmez** [comment]
Used prompt `python -m mlx_lm.generate --model /Users/gokdenizgulmez/Library/Mobile\ Documents/com\~apple\~CloudDocs/Datastes/MLX/OLMoE-1B-7B-0125-Instruct-4bit --prompt "what's your name"`

Before:
```text
Prompt: 17 tokens, 25.602 tokens-per-sec
Generation: 100 tokens, 4.439 tokens-per-sec
Peak memory: 3.939 GB
``

After:
```text
Prompt: 17 tokens, 79.650 tokens-per-sec
Generation: 100 tokens, 23.197 tokens-per-sec
Peak memory: 3.940 GB
```

File: llms/mlx_lm/models/olmoe.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+        B, L, D = x.shape
+        x = x.reshape(-1, D)
+        router_logits = self.gate(x)
+        routing_weights = mx.softmax(router_logits, axis=1, precise=True)
+        final_hidden_states = mx.zeros_like(x)
+        for expert_idx in range(self.num_experts):
+            expert_weights = routing_weights[:, expert_idx:expert_idx+1]
+            expert_output = self.experts[expert_idx](x)
+            final_hidden_states += expert_output * expert_weights
+        return final_hidden_states.reshape(B, L, D)
```
Comment (awni) [inline llms/mlx_lm/models/olmoe.py:125]: This whole thing should be replaced with a `SwitchGLU` layer. It will be much faster. See the QwenMOE implementation as an example: https://github.com/ml-explore/mlx-examples/blob/main/llms/mlx_lm/models/qwen2_moe.py#L110-L145

Note it will also require a `sanitize` function to pre-pack the weights.

**Comment by awni** [comment]
Thanks for adding that! Looks great overall except the one comment re using `SwitchGLU`. Please make the change and then we can merge it!

File: llms/mlx_lm/models/olmoe.py (hunk 1 above)
Comment (Goekdeniz-Guelmez) [reply llms/mlx_lm/models/olmoe.py]: Thanks! it's a lot faster now

**Comment by awni** [comment]
Do you mind to run the format hooks?

**Comment by Goekdeniz-Guelmez** [comment]
Sure I’ll run the format hooks when I’m home

**Review by awni — APPROVED** [review]
Thanks for the addition!!
